syntax = "proto3";
import "google/protobuf/timestamp.proto";

package ecloud;

message Empty {}

message Timestamps {
  int32 vehicle_index = 1; // vehicle_uid
  google.protobuf.Timestamp sm_start_tstamp = 2;
  google.protobuf.Timestamp client_start_tstamp = 3; 
  google.protobuf.Timestamp client_end_tstamp = 4;
}

message Ping {
  int32 tick_id = 1; // use as bool for server
  Command command = 2;
  repeated Timestamps timestamps = 3;
}

enum State {
  UNDEFINED = 0;
  NEW = 1;
  START = 2;
  ACTIVE = 3;
  ENDED = 4;
}

enum Command {
  TICK = 0;
  END = 1;
  REQUEST_DEBUG_INFO = 2;
}

enum VehicleState {
  REGISTERING = 0;
  CARLA_UPDATE = 1;
  UPDATE_INFO_OK = 2;
  GET_DESTINATION = 3;
  TICK_OK = 4; // regular OK / ack
  TICK_DONE = 5; // simulation ended --> include all Debug Info with this update
  OK = 6;
  ERROR = 7;
  DEBUG_INFO_UPDATE = 8;
}

message SimulationState {
  State state = 1;
  int32 tick_id = 2;
  int32 vehicle_index = 3; // vehicle_uid
  string test_scenario = 4;
  string application = 5;
  string version = 6;
  Command command = 7;
  repeated WaypointBuffer all_waypoint_buffers = 8;
  google.protobuf.Timestamp sm_start_tstamp = 9;
  int32 last_world_tick_time_ms = 10;
  bool is_edge = 11;
}

message WaypointRequest {
  int32 vehicle_index = 1;
}

message EdgeWaypoints {
  repeated WaypointBuffer all_waypoint_buffers = 1;
}

message WaypointBuffer {
  int32 vehicle_index = 1; // vehicle_uid
  repeated Waypoint waypoint_buffer = 2;
}

message Waypoint {
  string id = 1;
  Transform transform = 2;
  int32 road_id = 3;
  int32 section_id = 4;
  int32 lane_id = 5;
  float s = 6;
  bool is_junction = 7;
  float lane_width = 8;
  int32 lane_change = 9; // unused - enum if needed
  int32 lane_type = 10; // unused - enum if needed
  int32 right_lane_marking = 11; // unused - enum if needed
  int32 left_lane_marking = 12; // unused - enum if needed
}

message Transform {
  Location location = 1;
  Rotation rotation = 2;
}

message Location {
  float x = 1;
  float y = 2;
  float z = 3;
}

message Rotation {
  float yaw = 1;
  float pitch = 2;
  float roll = 3;
}

message Velocity {
  float x = 1;
  float y = 2;
  float z = 3;
}

message LocDebugHelper {
    repeated float gnss_x = 1;
    repeated float gnss_y = 2;
    repeated float gnss_yaw = 3;
    repeated float gnss_spd = 4;

    repeated float filter_x = 5;
    repeated float filter_y = 6;
    repeated float filter_yaw = 7;
    repeated float filter_spd = 8;

    repeated float gt_x = 9;
    repeated float gt_y = 10;
    repeated float gt_yaw= 11;
    repeated float gt_spd = 12;
}

message AgentDebugHelper {
    repeated float time_list = 1;
}

message PlanerDebugHelper {
    repeated float speed_list = 1;
    repeated float acc_list = 2;
    repeated float ttc_list = 3;
    repeated AgentDebugHelper agent_step_list = 4;
}

message ClientDebugHelper {
    repeated float perception_time_list = 1;
    repeated float localization_time_list = 2;
    repeated float update_info_time_list = 3;
    repeated float agent_update_info_time_list = 4;
    repeated float controller_update_info_time_list = 5;
    repeated float agent_step_time_list = 6;
    repeated float vehicle_step_time_list = 7;
    repeated float controller_step_time_list = 8;
    repeated float control_time_list = 9;
}

message VehicleUpdate {
  int32 tick_id = 1;
  int32 vehicle_index = 2;
  string vid = 3;
  int32 actor_id = 4;
  VehicleState vehicle_state = 5;
  LocDebugHelper loc_debug_helper = 6;
  PlanerDebugHelper planer_debug_helper = 7;
  ClientDebugHelper client_debug_helper = 8;
  google.protobuf.Timestamp sm_start_tstamp = 9;
  google.protobuf.Timestamp client_start_tstamp = 10; 
  google.protobuf.Timestamp client_end_tstamp = 11;
  float step_time_ms = 12;
  string container_name = 13;
  Transform transform = 14;
  Velocity velocity = 15;
}

message EcloudResponse {
    int32 tick_id = 1;
    repeated VehicleUpdate vehicle_update = 2;
}

service Ecloud {
  rpc Client_SendUpdate (VehicleUpdate) returns (SimulationState);
  rpc Client_RegisterVehicle (VehicleUpdate) returns (SimulationState);
  rpc Client_Ping (Ping) returns (Ping);
  rpc Client_GetWaypoints(WaypointRequest) returns (WaypointBuffer);
  // SERVER
  rpc Server_DoTick(SimulationState) returns (EcloudResponse);
  rpc Server_StartScenario(SimulationState) returns (EcloudResponse);
  rpc Server_EndScenario(Empty) returns (Empty);
  rpc Server_Ping(Empty) returns (Ping);
  rpc Server_GetVehicleUpdates(Empty) returns (EcloudResponse);
  rpc Server_PushEdgeWaypoints(EdgeWaypoints) returns (Empty);
}
